src/query/sql/src/planner/optimizer/decorrelate/subquery_rewriter.rs

```diff
@@ -453,6 +462,199 @@ impl SubqueryRewriter {
+        let (Some(left_source_binding), Some(right_source_binding)) = (
+            left_column.column.as_source(),
+            right_expr_binding.as_source(),
+        ) else {
+            return Ok(None);
+        };
+        // when left is a join, filtering becomes complicated. If in subquery is eliminated,
+        // the filter will be pushed down to Scan before join, resulting in inaccurate data results.
+        if left.has_join() {
+            return Ok(None);
+        }
+        if left_source_binding != right_source_binding {
+            return Ok(None);
+        }
+        let new_column_bindings = {
+            let guard = self.metadata.read();
+
+            let left_columns = guard.columns_by_table_index(left_table_index);
+            let right_columns = guard.columns_by_table_index(right_table_index);
+            let left_table = guard.table(left_table_index);
+            let right_table = guard.table(right_table_index);
+            // filter table function
+            if left_table.database() == "system" || right_table.database() == "system" {
+                return Ok(None);
+            }
+            let left_source_table_index = guard
+                .get_source_table_index(Some(left_table.database()), left_table.table().name());
+
+            if left_columns.len() != right_columns.len() {
+                return Ok(None);
+            }
+            let mut new_column_bindings = HashMap::with_capacity(left_columns.len());
+            for (left_entry, right_entry) in left_columns.into_iter().zip(right_columns.into_iter())
+            {
+                let (
+                    ColumnEntry::BaseTableColumn(left_column),
+                    ColumnEntry::BaseTableColumn(right_column),
+                ) = (left_entry, right_entry)
+                else {
+                    return Ok(None);
+                };
+                new_column_bindings.insert(
+                    right_column.column_index,
+                    ColumnBindingBuilder::new(
+                        left_column.column_name,
+                        left_column.column_index,
+                        Box::new(DataType::from(&left_column.data_type)),
+                        Visibility::Visible,
+                    )
+                    .table_name(Some(left_table.table().name().to_string()))
+                    .database_name(Some(left_table.database().to_string()))
+                    .table_index(Some(left_table_index))
+                    .source_table_index(left_source_table_index)
+                    .column_position(left_column.column_position)
+                    .virtual_expr(left_column.virtual_expr)
+                    .build(),
+                );
+            }
+            new_column_bindings
+        };
+        // restore ColumnBinding of same table in Subquery to table in Left
+        let mut body_visitor = SubqueryBodyVisitor::new(new_column_bindings, left.clone());
```

Comment:
we should rewrite the subquery child plan to replace equivalent columns, not the input `s_expr`.